src/sp/it/pl/core/CoreMenus.kt

```diff
@@ -51,177 +49,134 @@ class CoreMenus: Core {
+                            { it.name },
+                            {
+                                try {
+                                    it(selected)
+                                } catch (e: IllegalAccessException) {
+                                    logger.error(e) { "Could not invoke method $it on object $selected" }
+                                } catch (e: InvocationTargetException) {
+                                    logger.error(e) { "Could not invoke method $it on object $selected" }
+                                }
+                            }
+                    )
+                }
+        }
+        add<File> {
+            if (selected.isPlayable) {
+                item("Play") { PlaylistManager.use { it.playUri(selected.toURI()) } }
+                item("Enqueue") { PlaylistManager.use { it.addFile(selected) } }
+            }
+            item("Browse location") { selected.browse() }
+            item("Open (in associated program)") { selected.open() }
+            item("Edit (in associated editor)") { selected.edit() }
+            item("Delete from disc") { selected.recycle() }
+            item("Copy as ...") {
+                saveFile("Copy as...", APP.DIR_APP, selected.name, contextMenu.ownerWindow, ImageFileFormat.filter())
+                        .ifOk { nf ->
+                            // TODO: use customization popup
+                            val success = selected.copyRecursively(nf, false) { f, e ->
+                                logger.error(e) { "File copy failed" }
+                                OnErrorAction.SKIP
+                            }
+                            if (!success) APP.messagePane.show("File $selected copy failed")
                         }
-                    }
-            )
+            }
         }
-        add<File> { contextMenu, file ->
-            menuItems(
-                    item("Browse location") { file.browse() },
-                    item("Open (in associated program)") { file.open() },
-                    item("Edit (in associated editor)") { file.edit() },
-                    item("Delete from disc") { file.recycle() },
-                    item("Copy as ...") {
-                        saveFile("Copy as...", APP.DIR_APP, file.name, contextMenu.ownerWindow, ImageFileFormat.filter())
-                                .ifOk { nf ->
-                                    // TODO: use customization popup
-                                    val success = file.copyRecursively(nf, false) { f, e ->
-                                        logger.error(e) { "File copy failed" }
-                                        OnErrorAction.SKIP
-                                    }
-                                    if (!success) APP.messagePane.show("File $file copy failed")
-                                }
-                    }
-            )
+        addMany<File> {
+            if (selected.all { it.isPlayable }) {
```

Comment:
I suggest removing this menu.

1 This may have performance implications. isPlayable check may involve heavy-weight I/O, currently it does not, but it may. This is not invoked when user clicks on the menuItem to play the files, but when the context menu is being displayed. Which means the displaying could be delayed or blocked. We should be careful with i/o on ui thread.

2 You used files.all(isPlayable), but nothing prevented you to use files.any(isPlayable). Which would imply 'play any playable file in the set of files.' It is best to avoid logic that has multiple valid alternatives. Basically this action makes no sense, either both should be defined or none.

3 Generally conditional items signify a conceptual problem. This menu generation relies on type of the value. I.e. every type has its menu and upon displaying the menu for certain value, its type will determine which menu is shown. File is not something that is playable, it is different domain and you are crossing the FIle's responsibility here. It is Song, which should have a menu with play item, not File.

Now I understand the intended convenience of this, but it only makes sense in very loosely on an intuitive level. The correct menu item is:

```
add<File> {
    item("Play any playable") {  ...playlist.add(selected.filter(isPlayable)) }
}
```
1 Notice how all the computation (and I/O) is done in the action lambda, which happens after menu is show and can be easily made to run on non ui thread.
2 While File has no concept of being playable, this approach is fine, because we are in a media player application, so it is expected to consider an option of file being playable. But this way we do it without corrupting the File's purpose.

Persnally, I would remove the menuItem altogether, without the suggested substitute (I would retain the single File play menuItem). As for why: There is already ```item("Show detail") { APP.actionPane.show(o) }```, which allows the play functionality and does that with full user control over what to do.

Same applies for "enqueue".

Well, personally, I woul in the end not introduce the single file version either (but I can see how they would be more intuitive than going through "inspect". The reason being that there may be multiple playlists open and at that point play/enqueue also lose their exact definition. The inspect approach also resolves this issue completely.

Reply:
PS: the any {} vs all {} is what I call the 'AND vs OR' problem. Often user may desire the other behavior or is simply confused about the expected behavior. Usually bad design.

Reply:
Hmm I do get your point, but the route over "Show detail" is not intuitive. Also, if I click that on a FLAC, I only get offered "Open in converter" and "Set as data", nothing about play...

And I still would like to know where to test the addMany menu?

Reply:
I'll need to think about this some more and play around with the Show detail thing. I'm aware that it is far from perfect. Oh btw, if you are bothered by it covering the screen, Settings > View > Display method > Window.

For now at least move the computation from the predicate to the action, like I suggested. For multi file version.  It will cause the action to always appear but that is okey. Single file is better as is.